crates/storage/src/structure.rs

```diff
@@ -0,0 +1,132 @@
+/// the key and value and putting/extracting it to/from the storage.
+pub trait Structure<M, S>
+where
+    M: Mappable,
+    S: KeyValueStore,
+{
+    /// The codec used to encode and decode storage key.
+    type KeyCodec: Encode<M::Key> + Decode<M::OwnedKey>;
+    /// The codec used to encode and decode storage value.
+    type ValueCodec: Encode<M::Value> + Decode<M::OwnedValue>;
+
+    /// Puts the key-value pair into the storage.
+    fn put(
+        storage: &mut S,
+        key: &M::Key,
+        column: S::Column,
+        value: &M::Value,
+    ) -> StorageResult<()>;
+
+    /// Puts the key-value pair into the storage and returns the old value.
+    fn replace(
+        storage: &mut S,
+        key: &M::Key,
+        column: S::Column,
+        value: &M::Value,
+    ) -> StorageResult<Option<M::OwnedValue>>;
+
+    /// Takes the value from the storage and returns it.
+    /// The value is removed from the storage.
+    fn take(
+        storage: &mut S,
+        key: &M::Key,
+        column: S::Column,
+    ) -> StorageResult<Option<M::OwnedValue>>;
+
+    /// Removes the value from the storage.
+    fn delete(storage: &mut S, key: &M::Key, column: S::Column) -> StorageResult<()>;
+
+    /// Checks if the value exists in the storage.
+    fn exists(storage: &S, key: &M::Key, column: S::Column) -> StorageResult<bool> {
+        let key_encoder = Self::KeyCodec::encode(key);
+        let key_bytes = key_encoder.as_bytes();
+        storage.exists(key_bytes.as_ref(), column)
+    }
+
+    /// Returns the size of the value in the storage.
+    fn size_of_value(
+        storage: &S,
+        key: &M::Key,
+        column: S::Column,
+    ) -> StorageResult<Option<usize>> {
+        let key_encoder = Self::KeyCodec::encode(key);
+        let key_bytes = key_encoder.as_bytes();
+        storage.size_of_value(key_bytes.as_ref(), column)
+    }
+
+    /// Returns the value from the storage.
+    fn get(
+        storage: &S,
+        key: &M::Key,
+        column: S::Column,
+    ) -> StorageResult<Option<M::OwnedValue>> {
+        let key_encoder = Self::KeyCodec::encode(key);
+        let key_bytes = key_encoder.as_bytes();
+        storage
+            .get(key_bytes.as_ref(), column)?
+            .map(|value| {
+                Self::ValueCodec::decode_from_value(value).map_err(crate::Error::Codec)
+            })
+            .transpose()
+    }
+}
+
+/// It is an extension of the structure that allows supporting batch operations.
+/// Usually, they are more performant than initializing/inserting/removing values one by one.
+pub trait SupportsBatching<M, S>: Structure<M, S>
+where
+    M: Mappable,
+    S: BatchOperations,
+{
+    /// Initializes the storage with a bunch of key-value pairs.
+    /// In some cases, this method may be more performant than [`Self::insert`].
+    fn init(
+        storage: &mut S,
+        column: S::Column,
+        set: &mut dyn Iterator<Item = (&M::Key, &M::Value)>,
```

Comment:
I'd prefer a generic/`impl` here over type erasure. Is there a specific reason we want `dyn` over having some concrete type at compile time?